### PR TITLE
Ekeller/packaging

### PR DESCRIPTION
**Pull request by ethan92429**
Update to the OpenAPI converter. Make a bunch of changes to the grunt task to get the Endpoint definition.

### CODE REVIEW

File: Gruntfile.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                        const d = {groups: JSON.parse(data), version: version}
+                        fs.writeFile(opts.api_data_file_path, JSON.stringify(d), function (err) {
+                            if (err) {
+                                return console.log(err);
+                            }
+                            console.log("The file was saved!");
+                            resolve(d);
+                        });
+                    });
+                    // client.onshape.get({path: "/api/build"}, function (version) {
```
Comment (mthiesmeyer) [inline Gruntfile.js]: Is this commented section still necessary? Or can it be removed?

File: README.md [hunk 2]
```diff
@@ -1,2 +1,42 @@
 # api-generator
-Code to generate language specific interfaces based on Onshape apidoc.
+This repo is responsible for generating an OpenAPI specification for the Onshape API based on the JSON definition pulled from the Endpoints endpoint. During that process, it incorporates the various endpoint definitions hand written in OnshapeOpenApiSpecTemplate.yaml. 
+
+The most up-to-date version of the spec can always be found [here](https://github.com/onshape-public/api-generator/releases/latest).
+
+## Downloading a new API definition
+The API definition should be downloaded with the downloadApiDefinition grunt task, which can be run with:
+`grunt downloadApiDefinition`
+By defualt, this will place the downloaded file in the api_data folder with the name "apiData.json".
```
Comment (mthiesmeyer) [inline README.md]: Sp: default

File: README.md [hunk 3]
```diff
@@ -1,2 +1,42 @@
...
+
+## Running the parser
+To run the parser, update and run "python api_parser/json_to_openapi_2_0.py". So far, this command doesn't take command line arguments and you will therefore need to change the hardcoded values under the 'if __name__=="__main__": ' section at the end. There are a number of possible
```
Comment (mthiesmeyer) [inline README.md]: 'There are a number of possible ' ...?

File: api_parser/json_to_openapi_2_0.py [hunk 4]
```diff
@@ -0,0 +1,638 @@
...
+            if path in d['paths']:
+                Converter.nondestructive_update_dict(d['paths'][path], self.paths_dict[path])
+            else:
+                d['paths'][path] = self.paths_dict[path]
+        Converter.nondestructive_update_dict(d['definitions'], self.models_dict)
+        # Update the version
+        d["info"]["version"] = self.implementation_version
+        return d
+
+    # @staticmethod
```
Comment (mthiesmeyer) [inline api_parser/json_to_openapi_2_0.py]: Can be removed?

File: api_parser/json_to_openapi_2_0.py [hunk 5]
```diff
@@ -0,0 +1,638 @@
...
+            f['description'] = field['description']
+            fl.append(f)
+        return fl
+
+    @staticmethod
+    def convert_path(path):
+        """
+        "/accounts/:aid/purchases/:pid" -->  "/accounts/{aid}/purchases/{pid}"
+        """
+        return re.sub(r":(\w*)", r"{\1}", path)
```
Comment (mthiesmeyer) [inline api_parser/json_to_openapi_2_0.py]: Should we do '+' here (one or more)? As we should never have an empty /:/ url segment?

File: api_parser/json_to_openapi_2_0.py [hunk 6]
```diff
@@ -0,0 +1,638 @@
...
+        """Takes in an endpoint like "/assemblies/d/{did}/[wvm]/{wvm}/e/{eid}" and turns all the bracketed path
+        components, like [wvm] into a parameter definition that looks like:
+        -name: wvm
+         in: path
+         type: string
+         required: true
+         description: "One of "w" or "v" corresponding to whether a "workspace" or "version" was entered. """
+
+        # Get all items that match a bracketed string within the path
+        bracket_catcher = re.compile(r"\[(\w*)\]")
```
Comment (mthiesmeyer) [inline api_parser/json_to_openapi_2_0.py]: Similar to above here?

File: api_parser/json_to_openapi_2_0.py [hunk 7]
```diff
@@ -0,0 +1,638 @@
...
+        """Make a ref, like '#/definitions/Documents_getDocuments_Body'"""
+        return "#/definitions/{}".format(Converter.name_model(endpoint, location))
+
+
+if __name__ == "__main__":
+    converter = Converter(path='./api_data/apiData.json', template_path='./api_data/onshapeOpenApiSpecTemplate.yaml',
+                          config={'include_required': True, 'include_tags': True, 'inline_models': False})
+    yaml.dump(converter.converted_dict, open(converter.path + "Auto.yaml", "w"))
+
+    # converter = Converter(path='./api_data/apiDataAll.json', template_path='./api_data/onshapeOpenApiSpecTemplate.yaml',
```
Comment (mthiesmeyer) [inline api_parser/json_to_openapi_2_0.py]: Remove?

File: api_parser/test_converter.py [hunk 8]
```diff
@@ -0,0 +1,561 @@
...
+
+def test_add_required():
+    dic = {'a': {'b': {'c': 'c.a'}}}
+    keys = ['a', 'b']
+    value = "We need this!"
+    Converter.add_required(dic, keys, value)
+    assert dic == {'a': {'b': {'c': 'c.a', 'required': ['We need this!']}}}
+
+
+# def test_expand_dictionary():
```
Comment (mthiesmeyer) [inline api_parser/test_converter.py]: Remove?

File: generateApiWrapper.js [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+            resolve(opts);
+
+        }
+    );
+}
+
+function buildLanguageWrapper(opts) {
+    return new Promise(function (resolve, reject) {
+        switch (opts.language) {
+            case 'python':
```
Comment (mthiesmeyer) [inline generateApiWrapper.js]: Comment above states that C# is also allowed, but there is no case for it here?

File: generateApiWrapper.js [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    return new Promise(function (resolve, reject) {
+        optsCheck(opts).then(getEndpointsPromise).then(buildLanguageWrapper).then(resolve("success"));
+    })
+}
+
+module.exports = {
+    buildWrapper: buildWrapper
+};
+
+// buildWrapper(defaults).catch(function(data) {console.log(data)});
```
Comment (mthiesmeyer) [inline generateApiWrapper.js]: Remove?

File: tests/convert_response_list_to_definition.yaml [hunk 11]
```diff
@@ -0,0 +1,16 @@
+confiigurationParameters:
```
Comment (mthiesmeyer) [inline tests/convert_response_list_to_definition.yaml]: Is this double 'i' on purpose? I've seen it a few times throughout. If not, switch to 'configurationParameters'?

**Comment by mthiesmeyer** [comment]
Mostly nit/spelling comments, looks really good!

**Comment by ethan92429** [comment]
@mthiesmeyer just made the corrections. Thanks for the review!

**Comment by mthiesmeyer** [comment]
As this is a large review and will be public-facing code, please get at least one more pair of eyes on it :)

File: .gitignore [hunk 12]
```diff
@@ -0,0 +1,13 @@
+*Data.js
+*private/
+test.js
```
Comment (jramsley) [inline .gitignore]: test.js?  Why is this being ignored?

File: .gitignore [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+*.idea*
+__pycache__*
+_keys.py
+apikey.js
+node_modules
+python/generated
+.DS_Store
+*.onshapepy*
+api_data/apiData.json
```
Comment (jramsley) [inline .gitignore]: this is in gitignore but is also part of this commit.

File: Gruntfile.js [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+                client.getEndpoints(function (data) {
+                    console.log("Endpoint data received, first 100 chars of endpoint definition: " + data.slice(0, 100));
+                    request('https://cad.onshape.com/api/build', {json: true}, (err, res, version) => {
+                        if (err) {
+                            return console.log(err);
+                        }
+                        console.log("Version data recieved, implementation version is: " + version["Implementation-Version"]);
+                        const d = {groups: JSON.parse(data), version: version};
+                        ensureDirectoryExistence(opts.api_data_file_path);
+                        fs.writeFile(opts.api_data_file_path, JSON.stringify(d), function (err) {
```
Comment (jramsley) [inline Gruntfile.js]: Does this work?  I thought writeFile wanted a buffer.

File: Gruntfile.js [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+    api_keys: require('./apikey'),
+    language: 'Python',
+    includeInternalString: 'both',
+    data: [],
+    optionalParentDirectory: './generated'
+};
+
+function ensureDirectoryExistence(filePath) {
+    var dirname = path.dirname(filePath);
+    if (fs.existsSync(dirname)) {
```
Comment (jramsley) [inline Gruntfile.js]: This is called inside of a promise.  Mixing async with sync isn't a good idea.

**Pull request merged**